### PR TITLE
Added auto resource detector and exporter for GKE environment + relevant testing

### PR DESCRIPTION
**Pull request by clamclamyan**

### CODE REVIEW

**Comment by clamclamyan** [comment]
FYI @nilebox @jsuereth @punya

File: detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/EnvVarMock.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+import java.util.Map;
+
+public class EnvVarMock implements EnvVars {
```
Comment (nilebox) [inline detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/EnvVarMock.java:20]: This class can be moved to the `/test/` package I think

File: detectors/resources/src/test/java/com/google/cloud/opentelemetry/detectors/GKEResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            .willReturn(
+                aResponse().withHeader("Metadata-Flavor", "Google").withBody(responseBody)));
+  }
+
+  @Test
+  public void testGKEResourceNotGCP() {
+    GKEResource test = new GKEResource();
+
+    /* The default meta data url is unreachable through testing so getAttributes should not detect a
+    GCP environment, hence returning empty attributes */
```
Comment (nilebox) [inline detectors/resources/src/test/java/com/google/cloud/opentelemetry/detectors/GKEResourceTest.java:56]: This is correct for "GKE on GCP" (most common) but there are actually GKE distros to run on other platforms (Anthos), e.g. [GKE on AWS](https://cloud.google.com/anthos/clusters/docs/aws/concepts/architecture) or [GKE on-premise](https://cloud.google.com/anthos/clusters/docs/on-prem/1.3/overview).

Still, this is probably fine to merge as is, and add support for GKE outside GCP later.

File: examples/metrics/build.gradle [hunk 3]
```diff
@@ -15,8 +15,13 @@
  */
 plugins {
 	id 'java'
+	id 'com.google.cloud.tools.jib' version '2.7.1'
 }
 
+// Set jib image path up using Google Container Registry
```
Comment (nilebox) [inline examples/metrics/build.gradle:21]:
```suggestion
// Set Docker image path here, e.g. using Google Container Registry or Docker Hub
```

File: examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java [hunk 4]
```diff
@@ -18,17 +18,23 @@
...
 import java.io.IOException;
 import java.util.Random;
+import org.apache.log4j.BasicConfigurator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class MetricsExporterExample {
+  private static final Logger logger = LoggerFactory.getLogger(MetricsExporterExample.class);
+
+  private static SdkMeterProvider METER_PROVIDER =
```
Comment (nilebox) [inline examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java:34]: nit: Ideally these changes should be reviewed and merged separately, as they are not related to adding support for GKE resource.

But given that they are fixing issues with the example it might be okay to merge now and improve further later if needed.

File: exporters/metrics/src/main/java/com/google/cloud/opentelemetry/metric/MetricExporter.java [hunk 5]
```diff
@@ -167,7 +167,12 @@ public CompletableResultCode export(Collection<MetricData> metrics) {
           continue;
       }
       // Update metric descriptors based on configured strategy.
-      metricDescriptorStrategy.exportDescriptors(builder.getDescriptors(), this::exportDescriptor);
+      try {
```
Comment (nilebox) [inline exporters/metrics/src/main/java/com/google/cloud/opentelemetry/metric/MetricExporter.java:170]: This change is also not related to adding a GKE resource.
I think there is some issue with trying to create a metric descriptor for a metric which already exists in Cloud Monitoring, but not sure what is the correct way to handle this.

File: examples/metrics/build.gradle (hunk 3 above)
Comment (jsuereth) [inline examples/metrics/build.gradle:21]: You should call out this uses JIB in the README.md, and maybe a link to the gradle plugin docs: https://github.com/GoogleContainerTools/jib/tree/master/jib-gradle-plugin.

Ideally this would have a default setting for folks consuming the example.  Maybe put one in here for now with the doc line you list so folks know to go and change it?

Something generic like:

```
jib.to.image = 'gcr.io/my-gcp-project/my-app'
```

File: detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider [hunk 6]
```diff
@@ -1 +1,2 @@
-com.google.cloud.opentelemetry.detectors.GCEResource
\ No newline at end of file
+com.google.cloud.opentelemetry.detectors.GCEResource
+com.google.cloud.opentelemetry.detectors.GKEResource
```
Comment (jsuereth) [inline detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider:2]: How do these resource detects work?  Is it "last one wins"? Is it "all of them win"?

TL;DR: how do we ensure GKE beats GCE when it needs to?

**Review by jsuereth — CHANGES_REQUESTED** [review]

File: detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider (hunk 6 above)
Comment (nilebox) [reply detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider:2]: 1. It's "last write wins" in the current implementation: https://github.com/open-telemetry/opentelemetry-java/blob/bace5dce1f042d76ebbfec3e888f136361d58556/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#L182-L183
2. But I'm not sure if the order in the `META-INF` guarantees that classes are loaded in the same order (or if it's deterministic at all)? They are loaded using `ServiceLoader`: https://github.com/open-telemetry/opentelemetry-java/blob/bace5dce1f042d76ebbfec3e888f136361d58556/sdk/common/src/main/java/io/opentelemetry/sdk/resources/Resource.java#L105-L109

For now it shouldn't matter which one wins I think, but if we later add support for `cloud.infrastructure_service` (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md#cloud), we will need for `gcp_gke` to win over `gcp_gce`.

File: detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/EnvVars.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.opentelemetry.detectors;
+
+/** Provides API to fetch environment variables */
+public interface EnvVars {
+  EnvVars DEFAULT_INSTANCE = new EnvVarImpl();
```
Comment (punya) [inline detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/EnvVars.java:20]: Totally optional, but you can write this as

```java
EnvVars DEFAULT_INSTANCE = System::getenv;
```

and avoid creating a class.

File: detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GKEResource.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    if (containerName != null && !containerName.isEmpty()) {
+      attrBuilders.put(SemanticAttributes.K8S_CONTAINER_NAME, containerName);
+    }
+
+    String clusterName = metadata.getClusterName();
+    if (clusterName != null && !clusterName.isEmpty()) {
+      attrBuilders.put(SemanticAttributes.K8S_CLUSTER_NAME, clusterName);
+    }
+
+    attrBuilders.putAll(gceAttributes);
```
Comment (punya) [inline detectors/resources/src/main/java/com/google/cloud/opentelemetry/detectors/GKEResource.java:62]: Are `gceAttributes` meant to override the four things we pull from env variables above? I.e., would it be just as correct to move this code to line 47?

(You don't have to actually move the code, I'm asking just so we understand and document if there is an expected precedence order.)

File: detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider (hunk 6 above)
Comment (punya) [reply detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider:2]: Is the SPI going to be frozen after 1.0? I.e., will we be able to add a prioritization mechanism or will we have to couple/merge the GCE and GKE detectors to accomplish this goal?

File: examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java [hunk 9]
```diff
@@ -43,24 +49,24 @@ private static void setupMetricExporter() {
...
       throw new RuntimeException(e);
     }
   }
 
   private static void myUseCase() {
-    LongUpDownCounter counter =
+    LongCounter counter =
         METER
-            .longUpDownCounterBuilder("processed_jobs")
+            .longCounterBuilder("clamclamyan_example_counter")
```
Comment (punya) [inline examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java:62]: I would prefer if this username wasn't part of the code.

File: examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java [hunk 10]
```diff
@@ -72,13 +78,30 @@ private static void doWork(LongUpDownCounter counter) {
   }
 
   public static void main(String[] args) throws InterruptedException {
+    System.out.println("Starting the metrics-example application");
+
+    BasicConfigurator.configure();
+    logger.info("Test log message");
```
Comment (punya) [inline examples/metrics/src/main/java/com/google/cloud/opentelemetry/example/metrics/MetricsExporterExample.java:84]: Is the logger required? This code only appears to use `System.out.println`.

File: detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider (hunk 6 above)
Comment (jsuereth) [reply detectors/resources/src/main/resources/META-INF/services/io.opentelemetry.sdk.resources.ResourceProvider:2]: Not he SPI is not freezing with 1.0 yet, confirmed in the last insturmentation SIG.  I'm going to be making some contributions there, but I think we MAY want to iron out a test to make sure GCE/GKE detectors don't conflict.

@clamclamyan can you just open an Issue for us to figure this out later?  We don't need to hold up the PR on this, I was just curious how you were testing this locally and not running into issues.

**Comment by nilebox** [comment]
The `EndToEndTest` is failing for some reason, which doesn't seem related to changes in this PR:
```
org.testcontainers.containers.ContainerLaunchException: Container startup failed

        Caused by:
        org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
            at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
            at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323)
            ... 45 more

            Caused by:
            org.testcontainers.containers.ContainerLaunchException: Could not create/start container
                at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:497)
                at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325)
                at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
                ... 46 more

                Caused by:
                java.lang.IllegalStateException: Container did not start correctly.
                    at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:435)
                    ... 48 more
```

**Review by nilebox — APPROVED** [review]

**Comment by nilebox** [comment]
@jsuereth could you look into the issue with starting a container in EndToEndTest? It blocks this PR from merging.
